concrete/directed.go

```diff
@@ -76,18 +74,28 @@ func (g *DirectedGraph) AddNode(n graph.Node) {
 	g.maxID = max(g.maxID, n.ID())
 }
 
-func (g *DirectedGraph) AddDirectedEdge(e graph.Edge, cost float64) {
-	from, to := e.From(), e.To()
-	if !g.NodeExists(from) {
+func (g *DirectedGraph) SetEdge(e graph.Edge, cost float64) {
+	var (
+		from = e.From()
+		fid  = from.ID()
+		to   = e.To()
+		tid  = to.ID()
+	)
+
+	if fid == tid {
+		panic("concrete: adding self edge")
+	}
+
+	if !g.Has(from) {
 		g.AddNode(from)
```

Comment:
We don't mutate nodes, so yes it's the right thing to do.

Brendan has asked about a node mutation method and I'm in two minds about it. In the code I've written for D* Lite, I thought it might be helpful, but I didn't need it. If there is more pressure to add it, then we can. Currently, node mutation can only be done outside the API, by making sure the nodes are pointer types.